### ISSUE TITLE
你好，这个项目预训练和指令微调的代码，感觉似乎训练逻辑上没有什么区别，损失计算都是下一个词的交叉熵损失，数据处理也是完全一样

### ISSUE THREAD

**Issue by NICKQZL** [jingyaogong/minimind-v#67]
比较了下感觉好像这个项目预训练和指令微调的唯一不同之处就是预训练只训练 vision_proj的参数，而指令微调是全参数训练，但是数据上都是用的问答数据，训练方式也都是将问答数据用template转成字符串后，错位一下再自回归式预测下一个词的续写模式？那好像没啥区别，都是训练的文本的续写？而且都是 问答在被template转化后的文本的续写，没有本质区别？我是新手，看下来感觉所谓指令微调相较于预训练，想象中的针对 ”指令性“，”问答性“的微调过程 好像没有体现出来，依然像是预训练中的续写模式训练？

**Comment by jingyaogong** [comment]
现在的pretrain和instruct都是**指令训练**，取决于数据集

这里pretrain没有用大规模图文对，仅此而已

**Comment by NICKQZL** [comment]
> 现在的pretrain和instruct都是**指令训练**，取决于数据集
> 
> 这里pretrain没有用大规模图文对，仅此而已

好的，是指现在业内的大多数大模型的pretrain和sft阶段都是将问答数据 用template转文本后 训练续写吗？但是我疑惑的是为什么pretrain模型去回答问题会一直续写不知道啥时候停，但是sft模型就能知道啥时候该停，看起来明明训练的手法都一样的，本质都还是续写式训练（预测下一个词任务），数据也差不多都源自问答，是什么造成了两种模型一个不知道停一个知道要停

**Comment by jingyaogong** [comment]
都知道停的啊，有停止token </s>限制

**Comment by NICKQZL** [comment]
> 都知道停的啊，有停止token 限制

就是我指的不是靠max_length 强制停止，而是自然生成<eos>token停止。貌似我试了Qwen系列的基座模型（pretrain模型）自己生成<eos>停止是很难的，会一直说直到最长生成限制；而instruct版（sft模型）才能正常对话，输出答案后就立即输出<eos>停下

**Comment by xiao10ma** [comment]
我觉得是 pre-train 它只优化 vision mapping （CLIP dim -> hidden state dim） 这个 MLP，所以可能就会像单词接龙一样，一直说不停，只能靠 max_length 来结束。
而 Vision-SFT 阶段，它优化除 CLIP 以外所有的权重，因此能学习到怎么停止。
